### ISSUE TITLE
Rule: MEQP2.Classes.ConstructorOperations

### ISSUE THREAD

**Issue by schmengler** [extdn/extdn-phpcs#46]
### Rule
Only dependency assignment operations are allowed in constructor. No other operations are allowed.

The rule can be included from MEQP2 with severity 5 (default), and needs to be documented with examples.

### Reason
Magento instantiates all objects in the dependency graph early (if no proxy is used), so instantiation must 1. be fast and 2. not depend on any application state

**Comment by jissereitsma** [comment]
See https://github.com/extdn/extdn-phpcs/pull/54

**Comment by davidverholen** [comment]
@schmengler I totally agree, but(!) what about classes, that are not necessarily instantiated via the OM?

Value Objects for example should validate the input in the constructor (from what I've learned last year in amersfoort ;) ).

I think there is already a very similar discussion in the "when to use new" issue https://github.com/extdn/extdn-phpcs/issues/17

**Comment by jissereitsma** [comment]
I complete agree with @davidverholen . Classes like Value Objects should impose stricter rules upon constructor arguments than the ones supplied by PHP 7 (using type hinting with Strict Types enabled). Simple types (`double $moneyAmount`) are best converted into Value Objects (`Money $moneyAmount`) but still those Value Objects (`class Money`) should check whether their own arguments (`double $amount`) are valid - and in the example of this sentence, a double is not enough: The value could be zero (while maybe this makes less sense) or negative (which makes perhaps no sense).

But perhaps, we can negotiate the meaning of *dependency assignment operations* as the rule definition mentions it. Assuming that the variables `$foo` and `$bar` are injected through the constructor, I would say that the following needs to be taken into account:

- `$this->foo = $foo` and `$this->bar = $bar` are perfectly fine.
- `$this->setFoo($foo)` is essentially the same as `$this->foo = $foo`, except that using a separate method for this allows for better input validation (where PHP type hinting is not enough).
- `$this->foobar = $foo + $bar` is still an assignment, but I would say this is more configuration than assignment and it should not be allowed.
